Don't render carriage returns (\r). Fixes #2456

reyemxela opened the pull request:
<!-- If this is your first pull request for Fyne please read the contributor docs at:
https://github.com/fyne-io/fyne/wiki/Contributing.
Be sure that your work is based off `develop` branch. --> 

### Description:
<!-- A summary of the change included and which issue it addresses.
Please include any relevant motivation and background. -->
Changed both RichText (which most things use as a renderer) and TextGrid to ignore/remove \r characters from input.

Each fix seems to me like the most straight-forward approach, but if there are any tweaks I should make, please let me know.

Fixes #2456 

### Checklist:
<!-- Please tick these as appropriate using [x] -->

- [ ] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass. (some tests are failing in develop, but these changes don't seem to affect anything)

andydotxyz commented on widget/richtext.go:
```diff
@@ -359,6 +359,7 @@ func (t *RichText) updateRowBounds() {
 				continue
 			}
 			textSeg := seg.(*TextSegment)
+			textSeg.Text = strings.ReplaceAll(textSeg.Text, "\r", "") // remove "\r" characters that otherwise render as space
```
My understanding is that this comes as a result of loading files on (for example) windows right? So `SetText` may be a better place to do this (just a 1 time cost).

reyemxela replied:
That's a good point. I personally haven't experienced this issue, but you're right, a \r would probably only happen initially. I'll change that.

andydotxyz reviewed (changes requested):
Just a thought inline.
Plus this is a great thing to have a test for :)

reyemxela commented:
Moved checks to `SetText()` for labels, and `updateText()` for entries, as it's possible to paste carriage returns into an entry.

Also added basic tests for both cases.

As I was adding the tests, I realized that there are still ways this could get missed, such as setting `(label/entry).Text` directly. Honestly not sure how big of a deal that is.

andydotxyz commented:
> As I was adding the tests, I realized that there are still ways this could get missed, such as setting `(label/entry).Text` directly. Honestly not sure how big of a deal that is.

That is a very good point. Can we protect against that here just like SetText without too much more overhead? I think there is a check internally to see if the field was changed (as we have the caching internally)

reyemxela commented:
That's what I was trying to figure out, but didn't find it yet. I'll keep digging and mess with it.

andydotxyz commented:
> We need to make sure that this doesn't slow things down.

Yes we need to make sure it is not impacting performance. However the change is needed - loading a windows file will display blocks where the `\r` was.

reyemxela commented:
Definitely. I know it's a pretty rare edge case (it sounded like an easy fix when I saw the issue, so I thought it'd be a nice quick thing to tackle. lol).

As for the textgrid, there shouldn't be any noticeable impact. It's already looping through each character and handling spaces/tabs and other things, so it's just ignoring the \r there.

I was trying to simplify the Entry and Label fixes to only run when text is first set, but I'm running into other things there, including a potential bug in the label code. (if I can figure that out, I might make a separate PR for it)

A compromise could be to just leave the fix in the `SetText` functions and nowhere else, so it only runs once when actually setting the text. And if the user is trying to assign `(label/entry).Text` directly, they're responsible for cleaning up the text themselves.

andydotxyz commented:
> A compromise could be to just leave the fix in the `SetText` functions and nowhere else, so it only runs once when actually setting the text. And if the user is trying to assign `(label/entry).Text` directly, they're responsible for cleaning up the text themselves.

I'd rather not do this, in general the contract of `SetXxx(...)` is the same as `Xxx = ...; Refresh()`.

andydotxyz commented:
Is there an alternative approach that could take this forward? It would be good to get this in and keep the basic handling of fields vs Refresh() etc.

stuartmscott commented:
As this is a rendering issue, wouldn't it be better handled in the driver (maybe [here](https://github.com/fyne-io/fyne/blob/master/internal/painter/gl/gl_common.go#L88), and similar for software renderer)? `RenderedTextSize` would also need to ignore `\r`.

Consider someone making a notepad-style app, they open the app on Windows, load a file, type something, and then save it - only to find the app has stripped all the carriage returns from the file without warning and without an option to override this behaviour.

stuartmscott commented:
#2733
